File: 0001-two-sum/0001-two-sum.cpp
struct Node {
  int val;
  int id;
  bool operator<(const Node& other){
    return val < other.val;
  } 
  bool operator==(const Node& other) {
    return val == other.val;
  }
};

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
       std::vector<Node> nodes;
       for (int i = 0; i < nums.size(); ++i)
         nodes.push_back(Node{nums[i], i});
       int left = 0;
       int right = nodes.size() - 1;
       sort(nodes.begin(), nodes.end() );
       while (left < right) {
        while (nodes[left].val + nodes[right].val > target)
          --right;
        if (nodes[left].val + nodes[right].val == target) return {nodes[left].id , nodes[right].id};
        ++left;
      }
      return {};
    }
};

File: 0001-two-sum/0001-two-sum.rs
use std::collections::HashMap;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut map = HashMap::new();
        let mut res = vec![];
        for (i, n) in nums.iter().enumerate() {
            if let Some(&back_id) = map.get(&(target - n)) {
                res.push(back_id as i32);
                res.push(i as i32);
                return res;
            } 
            map.insert(n, i);
        }

        return res;
    }
}